lib/IDE/CompletionInstance.cpp

```diff
@@ -441,22 +439,21 @@ bool CompletionInstance::performCachedOperationIfPossible(
 
     // Create a new module and a source file using the current AST context.
     auto &Ctx = oldM->getASTContext();
-    auto *newM =
-        ModuleDecl::create(oldM->getName(), Ctx, oldM->getImplicitImportInfo());
+    auto *newM = ModuleDecl::createMainModule(Ctx, oldM->getName(),
+                                              oldM->getImplicitImportInfo());
```

Comment:
Can this be considered a hot path? We are adding up to the cache each time we retrieve the completion file for a replaced main module.

Reply:
We only perform a limited number of cached completions before throwing away the `CompilerInstance` (and therefore both the `ASTContext` and evaluator cache), so I don't think this should be an issue in practice. Note this doesn't just apply to the evaluator's cache, there are also a bunch of other allocations that we make per module and file (including the AST nodes themselves) that won't be deallocated until the `ASTContext` is torn down.

Eventually we'd like code completion to take full advantage of the request evaluator by having it invalidate certain parsing requests, which would invalidate any dependant requests and ensure we only need to re-compute the minimal amount depending on what was changed. Once we get to this point, we'll hopefully be able to tie memory lifetimes such that request invalidation causes the deallocation of any state previously computed.

Reply:
Very informative, thank you.